internal/rbd/replicationcontrollerserver.go

```diff
@@ -0,0 +1,349 @@
+func (cs *ControllerServer) EnableVolumeReplication(ctx context.Context,
+	req *replication.EnableVolumeReplicationRequest,
+) (*replication.EnableVolumeReplicationResponse, error) {
+	var rbdVol = &rbdVolume{}
+	var err error
+	// ger rbdVolume from VolumeID
+	rbdVol, err = cs.getVolumeFromID(ctx, req.GetVolumeId(), req.GetSecrets())
+	defer cs.cleanup(rbdVol)
+	if err != nil {
+		return nil, err
+	}
+	// extract the mirroring mode
+	mirroringMode, err := getMirroringMode(ctx, req.GetParameters())
+	if err != nil {
+		return nil, err
+	}
+
+	mirroringInfo, err := rbdVol.getImageMirroringInfo(ctx)
+	if err != nil {
+		err = fmt.Errorf("failed to get mirroring info of %s with error %s", rbdVol.String(), err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if mirroringInfo.State != librbd.MirrorImageEnabled {
+		err = rbdVol.enableImageMirroring(ctx, mirroringMode)
+		if err != nil {
+			util.ErrorLog(ctx, "failed to enable mirroring of %s with error %s", err.Error())
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+	}
+	return &replication.EnableVolumeReplicationResponse{}, nil
+}
+
+// DisableVolumeReplication disables replication on volume.
+func (cs *ControllerServer) DisableVolumeReplication(ctx context.Context,
+	req *replication.DisableVolumeReplicationRequest,
+) (*replication.DisableVolumeReplicationResponse, error) {
+	var rbdVol = &rbdVolume{}
+	var err error
+	// ger rbdVolume from VolumeID
+	rbdVol, err = cs.getVolumeFromID(ctx, req.GetVolumeId(), req.GetSecrets())
+	defer cs.cleanup(rbdVol)
+	if err != nil {
+		return nil, err
+	}
+
+	// extract the force option
+	force, err := getForceOption(ctx, req.GetParameters())
+	if err != nil {
+		return nil, err
+	}
+
+	mirroringInfo, err := rbdVol.getImageMirroringInfo(ctx)
+	if err != nil {
+		err = fmt.Errorf("failed to get mirroring info of %s with error %s", rbdVol.String(), err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if mirroringInfo.State == librbd.MirrorImageEnabled {
+		err = rbdVol.disableImageMirroring(ctx, force)
+		if err != nil {
+			util.ErrorLog(ctx, "failed to enable mirroring of %s with error %s", err.Error())
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+		// TODO check do we need to return abort after disabling the mirroring
+		return &replication.DisableVolumeReplicationResponse{}, nil
+	}
+
+	if mirroringInfo.State != librbd.MirrorImageDisabled {
+		return nil, status.Errorf(codes.Aborted, "Image is in %d Mode", mirroringInfo.State)
+	}
+
+	return &replication.DisableVolumeReplicationResponse{}, nil
+}
+
+// PromoteVolume promotes the volume as primary.
+func (cs *ControllerServer) PromoteVolume(ctx context.Context,
+	req *replication.PromoteVolumeRequest,
+) (*replication.PromoteVolumeResponse, error) {
+	var rbdVol = &rbdVolume{}
+	var err error
+	// ger rbdVolume from VolumeID
+	rbdVol, err = cs.getVolumeFromID(ctx, req.GetVolumeId(), req.GetSecrets())
+	defer cs.cleanup(rbdVol)
```

Comment:
Protect against parallel operations, because we can, and possibly safer to let one operation on the volume through at a time.

Reply:
the lock is taken in the getVolumeFromID function and will be released on the cleanup defer function.  Anything missing here

Reply:
Ack! did not look into that function, if that is done then we are good w.r.t serialization of requests.